src/worker/multithreaded.rs

```diff
@@ -4,34 +4,48 @@
 //! and allow scaling to more cores.
 //! However, that would be more complex, so we use this as a starting point.
 
-use std::{mem, io, sync::mpsc::{self, Sender}};
+use std::{mem, sync::mpsc::{self, Receiver, Sender}};
 use crate::decoder::MAX_COMPONENTS;
 use crate::error::Result;
 use super::{RowData, Worker};
 use super::immediate::ImmediateWorker;
 
+pub fn with_multithreading<T>(f: impl FnOnce(&mut dyn Worker) -> T) -> T {
+    #[cfg(not(feature = "rayon"))]
+    return self::enter_threads(f);
+
+    #[cfg(feature = "rayon")]
+    return jpeg_rayon::enter(|mut worker| {
+        f(&mut worker)
+    });
+}
+
 enum WorkerMsg {
     Start(RowData),
     AppendRow(Vec<i16>),
     GetResult(Sender<Vec<u8>>),
 }
-pub struct MultiThreadedWorker {
+
+#[derive(Default)]
+pub struct MpscWorker {
     senders: [Option<Sender<WorkerMsg>>; MAX_COMPONENTS]
 }
 
-impl Worker for MultiThreadedWorker {
-    fn new() -> Result<Self> {
-        Ok(MultiThreadedWorker {
-            senders: [None, None, None, None]
-        })
-    }
-    fn start(&mut self, row_data: RowData) -> Result<()> {
+pub struct StdThreadWorker(MpscWorker);
+
+impl MpscWorker {
+    fn start_with(
+        &mut self,
+        row_data: RowData,
+        spawn_worker: impl FnOnce(usize) -> Result<Sender<WorkerMsg>>,
+    ) -> Result<()> {
```

Comment:
This whole function could just be:
```rust
// if there is no worker thread for this component yet, start one
if self.senders[row_data.index].is_none() {
    self.senders[row_data.index] = Some(spawn_worker_thread(row_data.index)?);
}
self.senders[row_data.index]
    .as_mut()
    .unwrap()
    .send(WorkerMsg::Start(row_data))
    .expect("jpeg-decoder worker thread error");
Ok(())
```

Reply:
It may well be possible to utilize `Option::get_or_insert_with` as well, to remove that explicit `unwrap`. I just didn't want to mix refactor and function change, but I'll take a look.

Reply:
Understood. I was mostly pointing this out because I was confused why the code was so complicated when I read it the first time.